### PR TITLE
Abort test if the blueprint construction failed

### LINKED ISSUE #41: Error handling in constructHomeserver could be unified
Considering this code (with 2 places where an `err != nil` can happen):

https://github.com/matrix-org/complement/blob/9acaf3cd28e2d32024c80a3f04105e2cce7ff1e3/internal/docker/builder.go#L269-L284

If the deploy fails (first `err != nil`) it gets logged to the standard output via `log.Printf` (and `printLogs` is called explicitely).
If the runner fails, it gets logged to the Builder (and `printLogs` is called by the parent).

I think those 2 error logging should be unified: what do you think is the best approach?

I would recommend having some logging to the standard output, since it can impact the rest of the test.

For instance adding an `err` param to `printLogs`, which will get printed at the end of the server log if not `nil`, called inside `construct`.

I would be glad to make a PR for this.

---

Background info: I am trying to understand how #39 can be fixed, and I discovered that the Blueprints has an error (but only after spending a couple of hours and finally adding `COMPLEMENT_DEBUG=1` :)
```
2020/11/17 22:57:42 federation_one_to_one_room.hs2 : failed to run instructions: federation_one_to_one_room.hs2 : request http://localhost:32902/_matrix/client/r0/join/%21IiiKfQGoJGEqTPCTXN:hs1?access_token=MDAxMWxvY2F0aW9uIGhzMgowMDEzaWRlbnRpZmllciBrZXkKMDAxMGNpZCBnZW4gPSAxCjAwMWJjaWQgdXNlcl9pZCA9IEBib2I6aHMyCjAwMTZjaWQgdHlwZSA9IGFjY2VzcwowMDIxY2lkIG5vbmNlID0gbjNjbkNJeUs7bmxhMFBubAowMDJmc2lnbmF0dXJlIGjnOvc29ANDh6mwUukBi_1NocFirh6_gE1ZoogwQotxCg returned HTTP 404 Not Found : {"errcode":"M_UNKNOWN","error":"No known servers"}
```

### PR DESCRIPTION
**Pull request by oliverpool**
Closes #41

Since multiple errors can happen when constructing the blueprint, the `ConstructBlueprints` will log all of them and return the first.

### CODE REVIEW

File: internal/docker/builder.go [hunk 1]
```diff
@@ -179,12 +178,23 @@ func (d *Builder) ConstructBlueprintsIfNotExist(bs []b.Blueprint) error {
...
 	for _, bprint := range bs {
-		go d.construct(bprint, &bpWg)
+		go (func(bprint b.Blueprint) {
+			errc <- d.construct(bprint)
+		})(bprint)
 	}
-	bpWg.Wait()
+	var errs []error
+	for i := 0; i < len(bs); i++ {
+		errs = append(errs, <-errc...)
```
Comment (Kegsay) [inline internal/docker/builder.go:189]: We don't ever close `errc` so we'll leak resources. Call `close(errc)` after this `for` loop.

File: internal/docker/builder.go [hunk 2]
```diff
@@ -179,12 +178,23 @@ func (d *Builder) ConstructBlueprintsIfNotExist(bs []b.Blueprint) error {
...
+	}
+	if len(errs) > 0 {
```
Comment (Kegsay) [inline internal/docker/builder.go:191]: Is this really the right check? Won't `errs` just be `[nil, nil, nil, ...]` in the happy path?

File: internal/docker/builder.go [hunk 3]
```diff
@@ -211,18 +221,18 @@ func (d *Builder) ConstructBlueprints(bs []b.Blueprint) error {
...
+	if err != nil {
+		return []error{err}
 	}
 
 	runner := instruction.NewRunner(bprint.Name, d.debugLogging)
 	results := make([]result, len(bprint.Homeservers))
 	for i, hs := range bprint.Homeservers {
 		res := d.constructHomeserver(bprint.Name, runner, hs, networkID)
-		if res.err != nil && res.containerID != "" {
+		if res.err != nil {
```
Comment (Kegsay) [inline internal/docker/builder.go:234]: We still presumably want to check `if res.containerID != ""` before calling `printLogs` else if `constructHomeserver` failed before the container was made, there will be no logs to print.

File: internal/docker/builder.go [hunk 4]
```diff
@@ -211,18 +221,18 @@ func (d *Builder) ConstructBlueprints(bs []b.Blueprint) error {
 }
 
 // construct all Homeservers sequentially then commits them
-func (d *Builder) construct(bprint b.Blueprint, bpWg *sync.WaitGroup) {
-	defer bpWg.Done()
-	networkID := createNetwork(d.Docker, bprint.Name)
-	if networkID == "" {
-		return
+func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
```
Comment (Kegsay) [inline internal/docker/builder.go:224]: I'd just have `(err error)` as the response and fail-fast and bail out of this function the moment anything goes wrong. We don't need to aggregate errors from different homeserver images.

File: internal/docker/builder.go [hunk 5]
```diff
@@ -468,13 +484,18 @@ func createNetwork(docker *client.Client, blueprintName string) (networkID strin
 		},
 	})
 	if err != nil {
-		log.Printf("Failed to create docker network: %s\n", err)
-		return ""
+		return "", fmt.Errorf("%s: failed to create docker network. %w", blueprintName, err)
 	}
 	if nw.Warning != "" {
+		if nw.ID == "" {
+			return "", fmt.Errorf("%s: fatal warning while creating docker network. %w", blueprintName, err)
```
Comment (Kegsay) [inline internal/docker/builder.go:491]: `err` will be `nil` here, you probably meant to log `nw.Warning` instead.

**Review by Kegsay — CHANGES_REQUESTED** [review]
Few things.

File: internal/docker/builder.go (hunk 1 above)
Comment (oliverpool) [reply internal/docker/builder.go:189]: Done!

File: internal/docker/builder.go (hunk 3 above)
Comment (oliverpool) [reply internal/docker/builder.go:235]: Done!

File: internal/docker/builder.go (hunk 5 above)
Comment (oliverpool) [reply internal/docker/builder.go]: Good catch!

File: internal/docker/builder.go (hunk 2 above)
Comment (oliverpool) [reply internal/docker/builder.go:192]: I am appending `<-errc...` to `errs`, so if the returned error array (from the channel) is empty, nothing will be appended and the length of `errs` will not change.

File: internal/docker/builder.go (hunk 4 above)
Comment (oliverpool) [reply internal/docker/builder.go:225]: I think it is better to try everything and then informing the user of all the failures (instead to have only one failure at a time, which can be very time consuming to debug).

Comment (oliverpool) [reply internal/docker/builder.go:225]: But if you prefer I can change this to fail early and ignore the rest.

Comment (Kegsay) [reply internal/docker/builder.go:225]: I don't have strong opinions either way. I generally lean on the side of "what makes the code simpler" than "what presents the most information to the user" when it comes to internal guts. For tests themselves then it's the other way around. I don't mind.

File: internal/docker/builder.go (hunk 2 above)
Comment (Kegsay) [reply internal/docker/builder.go:192]: It's not super clear this is the case. Please add a comment to the append.

**Review by Kegsay — CHANGES_REQUESTED** [review]
Please add the comment then LGTM! Thanks!

File: internal/docker/builder.go (hunk 2 above)
Comment (oliverpool) [reply internal/docker/builder.go:195]: I added a comment, is it explicit enough?

**Review by Kegsay — APPROVED** [review]

**Pull request merged**